src/main/java/org/jenkinsci/plugins/workflow/steps/EnvStep.java

```diff
@@ -150,11 +153,37 @@
                     }
                 }
                 return b.toString();
+            } else if (overrides instanceof Map) {
+                return ((Map<?, ?>) overrides).keySet()
+                    .stream()
+                    .filter(e -> e instanceof String)
+                    .map(Object::toString)
+                    .collect(Collectors.joining(", "));
             } else {
                 return null;
             }
         }
 
+        @NonNull
+        @Override
+        public Map<String, Object> customInstantiate(@NonNull Map<String, Object> arguments) {
+            Map<String, Object> r = new HashMap<>(arguments);
+            final String key = "overrides";
+            Object overrides = r.get(key);
+            if (overrides instanceof Map) {
+                r.put(key, toKeyValueList((Map<?, ?>) overrides));
+            }
+            return r;
+        }
+
+        private static List<String> toKeyValueList(Map<?, ?> map) {
+            return map.entrySet().stream()
+                .filter(e -> e.getKey() instanceof String)
```

Comment:
Clearer IMO to just cast to `String`, so that people will get a clear error if they use something else.
```suggestion
```